### DISCUSSION TITLE
Distance Calculation - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#7525] [Docs]
# Distance Calculation - Ultralytics YOLOv8 Docs

Distance Calculation Using Ultralytics YOLOv8

https://docs.ultralytics.com/guides/distance-calculation/

**Comment by sgjoonha** [comment]
how can i solve this problem?

ModuleNotFoundError: No module named 'ultralytics.solutions'

**Reply by RizwanMunawar** [reply]
@sgjoonha Try to upgrade the ultralytics package, hopefully, this will resolve your issue. i.e

```pip install -U ultralytics```

Thanks
Ultralytics Team!

**Reply by sgjoonha** [reply]
Thanks :)

**Reply by pderrenger** [reply]
@sgjoonha you're welcome! If you have any more questions or run into any issues, feel free to reach out. Happy coding with YOLOv8! 🚀😊

**Comment by MohamedAyman-991** [comment]
is it accurate enough for  distance calculation

**Reply by MohamedAyman-991** [reply]
and dose it work with segmented object ?

**Reply by RizwanMunawar** [reply]
@MohamedAyman-991 Yes, the distance calculation is ~accurate. Yes, it can work with segmented objects, but we have not officially released the support for this. We will add this support in the coming days.

Thanks
Ultralytics Team!

**Reply by MohamedAyman-991** [reply]
is it accurate for depth calculation i want to check if some object is inside some area

**Reply by RizwanMunawar** [reply]
Yes, It's accurate for distance calculation, not depth calculation.

Thanks
Ultralytics Team!

**Comment by TortoiseFather** [comment]
Unsure if I'm missing something but when running the example code using a custom video, clicking on the bounding boxes doesn't seem to do anything?

**Reply by TortoiseFather** [reply]
Also clicking on an area without a bounding box/attempting to draw a bounding box of my own also does nothing.

**Reply by RizwanMunawar** [reply]
@mylovelykohai Thanks for the update. I have tested the code and it's working fine for me. Make sure to use LEFT_MOUSE_CLICK to select two bounding boxes and then the code will start processing the videos (with distance calculation between objects). 

Thanks
Ultralytics Team!

**Reply by TortoiseFather** [reply]
It seems that it was an error regarding not using distance detection on first run-through
In case anyone has the same problem I had, (left mouse button not responding), delete the distance_calculation.avi and re-run code.

Unsure why it happened but I was tearing my hair out

**Reply by AdadAlShabab** [reply]
Can you please share the code. Please? It's still not working for me. @mylovelykohai

**Reply by TortoiseFather** [reply]
> Can you please share the code. Please? It's still not working for me. @mylovelykohai 

Delete the file distance_calculation.avi and then change the code to

video_writer = cv2.VideoWriter("MyFile.avi",
                               cv2.VideoWriter_fourcc(*'mp4v'),
                               fps,
                               (w, h))

Worked for me

**Reply by AdadAlShabab** [reply]
@mylovelykohai , still not working. 
This is the code:
from ultralytics import YOLO
from ultralytics.solutions import distance_calculation
import cv2

model = YOLO("yolov8n.pt")
names = model.model.names

cap = cv2.VideoCapture("E:/Ultralytics/file.mp4")
assert cap.isOpened(), "Error reading video file"
w, h, fps = (int(cap.get(x)) for x in (cv2.CAP_PROP_FRAME_WIDTH, cv2.CAP_PROP_FRAME_HEIGHT, cv2.CAP_PROP_FPS))

# Video writer
video_writer = cv2.VideoWriter("MyFile.avi",
                               cv2.VideoWriter_fourcc(*'mp4v'),
                               fps,
                               (w, h))

# Init distance-calculation obj
dist_obj = distance_calculation.DistanceCalculation()
dist_obj.set_args(names=names, view_img=True)

while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        print("Video frame is empty or video processing has been successfully completed.")
        break

    tracks = model.track(im0, persist=True, classes=0, show=True, tracker="bytetrack.yaml")
    im0 = dist_obj.start_process(im0, tracks)
    video_writer.write(im0)

cap.release()
video_writer.release()
cv2.destroyAllWindows()

**Reply by TortoiseFather** [reply]
Sorry, assuming file.mp4 exists and all the packages are installed that's the exact same code I've got

**Reply by pderrenger** [reply]
@mylovelykohai It seems like you've managed to resolve the issue by deleting the `distance_calculation.avi` file and rerunning the code. If you're still experiencing issues, ensure that your environment is set up correctly and that you're using the latest version of the Ultralytics YOLOv8 repository. If the problem persists, please provide more details about the error messages you're receiving, and we'll do our best to assist you further.

Thanks for your patience,
Ultralytics Team!

**Reply by TortoiseFather** [reply]
@pderrenger it's not that there's an error message, upon initial run the program was not recognising my mouse, but deleting the file fixed it, really not sure why.

**Reply by pderrenger** [reply]
@mylovelykohai I'm glad to hear that deleting the `distance_calculation.avi` file resolved the issue with mouse recognition. It's possible that there was a conflict or a lock on the file from a previous run that prevented the mouse events from being registered correctly. Starting fresh by removing the output file can often resolve such conflicts. If you encounter any further issues or have questions, feel free to reach out.

**Comment by jasminaaa20** [comment]
Hi. How to is distance calculated? Between the closest boundaries of the bounding boxes or between the center points of the bounding boxes?

**Reply by RizwanMunawar** [reply]
@jasminaaa20 The distance is computed based on the centroids of the two bounding boxes.

Thanks
Ultralytics Team!

**Reply by jasminaaa20** [reply]
@RizwanMunawar Thank you for the reply. Just a little suggestion. It would be great if there is a method to get the shortest distance between two objects, which is the distance between the closest boundaries of the bounding boxes (not considering 3 dimensional issues)

**Reply by RizwanMunawar** [reply]
@jasminaaa20, appreciate the suggestion. We continuously optimize modules based on user feedback. If you'd like to implement this feature in the meantime, feel free to open a pull request (PR). 

Thanks,
Ultralytics Team!

**Comment by taaha42777** [comment]
Hi, Yolo Team
I want to ask is their any possibility that i can use the distance calculation in real time via webcam, if yes kindly tell what commands, code to be added any tips or tricks also would be helpful. Actually i am using this yolo v8 framework in a project , I have also tried chat gpt 4 to try to convert into real time but no luck so far , please explain in detail.
kind regards

**Reply by taaha42777** [reply]
Fatal Python error: Aborted

Main thread:
Current thread 0x0000140c (most recent call first):
  File "C:\Users\taaha\AppData\Roaming\Python\Python311\site-packages\ultralytics\trackers\byte_tracker.py", line 95 in multi_gmc
  File "C:\Users\taaha\AppData\Roaming\Python\Python311\site-packages\ultralytics\trackers\byte_tracker.py", line 282 in update
  File "C:\Users\taaha\AppData\Roaming\Python\Python311\site-packages\ultralytics\trackers\track.py", line 64 in on_predict.postprocess_end
  File "C:\Users\taaha\AppData\Roaming\Python\Python311\site-packages\ultralytics\engine\predictor.py", line 401 in run_callbacks
  File "C:\Users\taaha\AppData\Roaming\Python\Python311\site-packages\ultralytics\engine\predictor.py", line 292 in stream_inference
  File "C:\Users\taaha\AppData\Roaming\Python\Python311\site-packages\torch\utils\_contextlib.py", line 35 in generator_context
  File "C:\Users\taaha\AppData\Roaming\Python\Python311\site-packages\ultralytics\engine\model.py", line 275 in predict
  File "C:\Users\taaha\AppData\Roaming\Python\Python311\site-packages\ultralytics\engine\model.py", line 296 in track
  File "C:\Users\taaha\.spyder-py3\temp.py", line 34 in <module>
  File "C:\ProgramData\anaconda3\Lib\site-packages\spyder_kernels\customize\spydercustomize.py", line 356 in compat_exec
  File "C:\ProgramData\anaconda3\Lib\site-packages\spyder_kernels\customize\spydercustomize.py", line 473 in exec_code
  File "C:\ProgramData\anaconda3\Lib\site-packages\spyder_kernels\customize\spydercustomize.py", line 615 in _exec_file
  File "C:\ProgramData\anaconda3\Lib\site-packages\spyder_kernels\customize\spydercustomize.py", line 528 in runfile
  File "C:\Users\taaha\AppData\Local\Temp\ipykernel_14536\1655235031.py", line 1 in <module> 






these type of error arrise as i run the above code, i have Lenovo G-50-80 core i 5 with 8 gb ram do i also need specific graphic card please explain in detail

**Reply by pderrenger** [reply]
@taaha42777 hello! It seems you're encountering a fatal Python error while running the Ultralytics YOLOv8 code. This type of error can be caused by various issues, including problems with the underlying C libraries, conflicts between packages, or hardware acceleration issues.

Regarding your first question about real-time distance calculation using a webcam with YOLOv8, yes, it is possible. The code snippet provided in the documentation you referenced is almost ready for webcam use. To adapt it for real-time webcam capture, you can simply change the `VideoCapture` source to `0` (which typically represents the default webcam on a system). Here's a modified version of the code for webcam use:

```python
from ultralytics import YOLO
from ultralytics.solutions import distance_calculation
import cv2

model = YOLO("yolov8n.pt")
names = model.names

# Use 0 for the default webcam
cap = cv2.VideoCapture(0)
assert cap.isOpened(), "Error accessing webcam"

# No need for video writer when using webcam in real-time
dist_obj = distance_calculation.DistanceCalculation()
dist_obj.set_args(names=names, view_img=True)

while True:
    success, im0 = cap.read()
    if not success:
        print("Error reading from webcam")
        break

    tracks = model.track(im0, persist=True, show=False)
    im0 = dist_obj.start_process(im0, tracks)
    cv2.imshow('Distance Calculation', im0)

    # Break the loop if 'q' is pressed
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()
```

Regarding the fatal Python error, it's difficult to diagnose the exact cause without more context, but here are some general tips:

1. **Update Packages**: Ensure that all your Python packages, especially `ultralytics`, are up to date.
2. **Environment**: Sometimes, package conflicts can cause issues. Consider creating a fresh Python environment using `conda` or `venv`.
3. **Hardware**: While a specific graphics card is not strictly necessary, using one that supports CUDA can significantly improve performance for YOLOv8. Your error does not seem to be directly related to the absence of a GPU, but performance will be limited on CPU-only setups.
4. **Debugging**: Try running your script line by line or with a debugger to pinpoint the exact line causing the issue.
5. **Permissions**: Ensure that the user running the script has the necessary permissions to access the webcam and other system resources.
6. **Spyder**: If you're using Spyder, try running the script from a terminal instead, as IDEs can sometimes cause unexpected behavior.

If you continue to experience issues, consider reaching out to the Ultralytics GitHub repository with detailed information about the error, your environment, and the steps leading up to the issue. The community or maintainers might be able to provide more targeted assistance.

**Comment by AdadAlShabab** [comment]
When running the example code using a custom video, clicking on the bounding boxes doesn't seem to do anything. Why? @RizwanMunawar

**Reply by RizwanMunawar** [reply]
@AdadAlShabab which platform you are using? Also, make sure to follow the mentioned guidelines.

- Mouse Right Click will delete all drawn points
- Mouse Left Click can be used to draw points

Hope this helps! Let me know if you have any additional questions.

Thanks and Regards
Ultralytics Team!

**Reply by AdadAlShabab** [reply]
@RizwanMunawar Brother, I followed the guideline of clicking mouse. But it isn't working. I'm running the code in Windows 11. The code is running perfectly but the mouse clicking & marking feature ain't working, which I needed. Please let me know, thanks!

**Comment by AdadAlShabab** [comment]
Hi @pderrenger . I'm using the following code:

`from ultralytics import YOLO
from ultralytics.solutions import distance_calculation
import cv2

model = YOLO("yolov8n.pt")
names = model.model.names

cap = cv2.VideoCapture("file.mp4")
assert cap.isOpened(), "Error reading video file"
w, h, fps = (int(cap.get(x)) for x in (cv2.CAP_PROP_FRAME_WIDTH, cv2.CAP_PROP_FRAME_HEIGHT, cv2.CAP_PROP_FPS))

# Video writer
video_writer = cv2.VideoWriter("MyFile.avi",
                               cv2.VideoWriter_fourcc(*'mp4v'),
                               fps,
                               (w, h))

# Init distance-calculation obj
dist_obj = distance_calculation.DistanceCalculation()
dist_obj.set_args(names=names, view_img=True)

while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        print("Video frame is empty or video processing has been successfully completed.")
        break

    tracks = model.track(im0, persist=True, show=False)
    im0 = dist_obj.start_process(im0, tracks)
    video_writer.write(im0)

cap.release()
video_writer.release()
cv2.destroyAllWindows()`

But the mouse clicking distance calculation feature ain't working. I've updated to latest Ultralytics too. I'm using Windows 11 machine. Please help. Thanks in advance.

**Reply by pderrenger** [reply]
@AdadAlShabab hi there! It seems like you're trying to use the distance calculation feature with YOLOv8. If the mouse clicking functionality isn't working as expected, here are a few things you can check:

1. **Ensure GUI Compatibility**: Make sure that your environment supports GUI operations. Some environments, especially remote servers or headless setups, may not support direct GUI interactions.

2. **Check OpenCV Version**: Ensure you have the correct version of OpenCV installed that supports GUI features. You can check your OpenCV version with `cv2.__version__`.

3. **Verify Callbacks**: The distance calculation feature relies on OpenCV's GUI callbacks. Make sure that the `distance_calculation` object is correctly handling mouse events.

4. **Review Code**: Double-check that the `view_img` argument is set to `True` in the `set_args` method, which is required to display the frames with the distance calculation.

5. **Update Packages**: Ensure all related packages are up to date, including `ultralytics`.

6. **Check for Errors**: Look for any error messages or warnings in the console that might indicate what the issue is.

7. **Run Minimal Example**: Try running a minimal example script to isolate the problem. Sometimes other parts of the code might interfere with the GUI functionality.

8. **Operating System Permissions**: Ensure that your application has the necessary permissions to create windows and capture mouse events on your Windows 11 machine.

If you've checked all of the above and the issue persists, it might be helpful to reach out to the Ultralytics GitHub issues page with detailed information about your setup and the problem you're encountering. The community and maintainers can provide more targeted assistance. Remember to include details about your environment, installed package versions, and any error messages you're receiving.

**Comment by dinesht27** [comment]
Hello, 

I already have implemented the yolo v8 inference for object detection, with onnxruntime, in c++ and the real time performance great. 

I would like to extend this to the Object tracking and Distance Estimation of the objects from the Camera. But I noticed there is already implementation of the botsort and bytetrack used in python from Ultralytics. is there a example c++ implementation or steps where I can extend this, 

Also I would as well further intend to extend the following possibilities, 

Object Blurring, 
Speed estimation, 
Counting, 
and as well the heatmaps images, for the same, 


I would appreciate if there is some reference, to get started soon, would appreciate this support much, Thank you

**Reply by pderrenger** [reply]
@dinesht27 hello,

Thank you for your interest in extending your YOLOv8 implementation to include object tracking and distance estimation! It's great to hear that you are achieving excellent real-time performance with ONNX Runtime in C++.

### Object Tracking and Distance Estimation

Currently, Ultralytics provides implementations for object tracking using BoT-SORT and ByteTrack in Python. While we don't have a direct C++ example, you can certainly adapt the existing Python logic to C++. Here are some steps to get you started:

1. **Object Tracking**: You can use the ONNX model for inference in C++ and implement the tracking logic based on the BoT-SORT or ByteTrack algorithms. You might find the [ByteTrack repository](https://github.com/ifzhang/ByteTrack) useful as a reference for the tracking algorithm.

2. **Distance Estimation**: For distance calculation, you can use the bounding box centroids. The `DistanceCalculation` class in Python can be translated to C++ by using OpenCV for drawing and calculating distances. Here's a simplified example in Python that you can adapt:

    ```python
    import cv2
    from ultralytics import YOLO, solutions

    model = YOLO("yolov8n.pt")
    names = model.model.names

    cap = cv2.VideoCapture("path/to/video/file.mp4")
    assert cap.isOpened(), "Error reading video file"
    w, h, fps = (int(cap.get(x)) for x in (cv2.CAP_PROP_FRAME_WIDTH, cv2.CAP_PROP_FRAME_HEIGHT, cv2.CAP_PROP_FPS))

    # Video writer
    video_writer = cv2.VideoWriter("distance_calculation.avi", cv2.VideoWriter_fourcc(*"mp4v"), fps, (w, h))

    # Init distance-calculation obj
    dist_obj = solutions.DistanceCalculation(names=names, view_img=True)

    while cap.isOpened():
        success, im0 = cap.read()
        if not success:
            print("Video frame is empty or video processing has been successfully completed.")
            break

        tracks = model.track(im0, persist=True, show=False)
        im0 = dist_obj.start_process(im0, tracks)
        video_writer.write(im0)

    cap.release()
    video_writer.release()
    cv2.destroyAllWindows()
    ```

### Additional Features

For the additional features you mentioned (Object Blurring, Speed Estimation, Counting, and Heatmaps), here are some brief pointers:

- **Object Blurring**: You can use OpenCV to blur regions of interest (ROIs) based on the bounding boxes.
- **Speed Estimation**: Track the movement of objects frame-by-frame and calculate speed based on the change in position over time.
- **Counting**: Maintain a count of unique object IDs detected over the frames.
- **Heatmaps**: Aggregate the positions of detected objects over time to generate heatmaps.

### Resources

For more detailed guidance on distance calculation, you can refer to our [Distance Calculation Guide](https://docs.ultralytics.com/guides/distance-calculation/).

Feel free to reach out if you have any more questions or need further assistance. Happy coding! 🚀
